Convert status in cluster-autoscaler-status to yaml and add error info for scale-up backoff

walidghallab opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR changes the status field in data of cluster-autoscaler-status config map to yaml to make it easier for parsing.
This PR also adds error info for scale-up during backoff to provide the reason of the backoff to the users to make it easier to debug.

#### Which issue(s) this PR fixes:

Fixes #6318

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
Changed status value in data of cluster-autoscaler-status config map to yaml format for easier parsing.

Added error information for backoff status in scale-up.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs

```

walidghallab commented:
/uncc @feiskyer 
/uncc @BigDarkClown

walidghallab commented:
/cc @towca 
/assign @towca

MaciekPytel reviewed (commented):
I'm not going to block review on this, since this is mostly the preexisting state. But the unittests for configmap are very old and, frankly, pretty terrible. I wonder if you wouldn't mind re-writing them into table-based tests? Your changes are completely re-working this part of the code and it's a good opportunity to refresh the tests. Also good test coverage would help make sure we're not introducing any regressions.

MaciekPytel commented:
Sorry, one more comment - would you mind posting an example yaml obtained via `kubectl describe configmap` or similar? Given that primary use-case is I think manual debugging, I think it's important to make sure the yaml is pretty-printed. Since your unittest does YAML comparison, I don't actually know if the real output is formated in similar way to your example file.

walidghallab commented on cluster-autoscaler/clusterstate/api/types.go:
```diff
@@ -69,36 +69,148 @@ const (
+}
+
+// HealthConditionNodeCount contains number of nodes that satisfy different criteria for health condition.
+type HealthConditionNodeCount struct {
```
Sounds good, Done.

walidghallab commented on cluster-autoscaler/clusterstate/api/types.go:
```diff
@@ -69,36 +69,148 @@ const (
+}
+
+// ClusterScaleUpNodeCount contains number of nodes that satisfy different criteria for scale up condition of the whole cluster.
+type ClusterScaleUpNodeCount struct {
```
I agree, Removed.

walidghallab commented on cluster-autoscaler/clusterstate/api/types.go:
```diff
@@ -69,36 +69,148 @@ const (
+}
+
+// NodeGroupScaleUpNodeCount contains number of nodes that satisfy different criteria for scale up condition of a node group.
+type NodeGroupScaleUpNodeCount struct {
```
Done

walidghallab commented on cluster-autoscaler/clusterstate/api/types.go:
```diff
@@ -69,36 +69,148 @@ const (
+	// Status of the scale up.
+	Status ClusterAutoscalerConditionStatus `json:"status,omitempty" yaml:"status,omitempty"`
+	// NodeCounts contains number of nodes that satisfy different criteria for scale up condition.
+	NodeCounts ClusterScaleUpNodeCount `json:"nodeCounts,omitempty" yaml:"nodeCounts,omitempty"`
```
Removed ClusterScaleUpNodeCount (but not Status)

walidghallab commented on cluster-autoscaler/clusterstate/api/types.go:
```diff
@@ -69,36 +69,148 @@ const (
+	// Status of the scale up.
+	Status ClusterAutoscalerConditionStatus `json:"status,omitempty" yaml:"status,omitempty"`
+	// NodeCounts contains number of nodes that satisfy different criteria for scale up condition.
+	NodeCounts NodeGroupScaleUpNodeCount `json:"nodeCounts,omitempty" yaml:"nodeCounts,omitempty"`
```
Removed NodeGroupScaleUpNodeCount (but not Status)

walidghallab commented on cluster-autoscaler/clusterstate/clusterstate.go:
```diff
@@ -757,42 +756,45 @@ func (csr *ClusterStateRegistry) UpdateScaleDownCandidates(nodes []*apiv1.Node,
 	result := &api.ClusterAutoscalerStatus{
-		ClusterwideConditions: make([]api.ClusterAutoscalerCondition, 0),
-		NodeGroupStatuses:     make([]api.NodeGroupStatus, 0),
+		AutoscalerStatus: "Running",
```
Good point, thank you!
Done.

walidghallab commented on cluster-autoscaler/clusterstate/clusterstate.go:
```diff
@@ -757,42 +756,45 @@ func (csr *ClusterStateRegistry) UpdateScaleDownCandidates(nodes []*apiv1.Node,
 	result := &api.ClusterAutoscalerStatus{
-		ClusterwideConditions: make([]api.ClusterAutoscalerCondition, 0),
-		NodeGroupStatuses:     make([]api.NodeGroupStatus, 0),
+		AutoscalerStatus: "Running",
```
Good point, done.

walidghallab commented on cluster-autoscaler/clusterstate/utils/status.go:
```diff
@@ -67,7 +69,7 @@ func NewStatusMapRecorder(kubeClient kube_client.Interface, namespace string, re
 	var err error
 	if active {
-		mapObj, err = WriteStatusConfigMap(kubeClient, namespace, "Initializing", nil, statusConfigMapName)
+		mapObj, err = WriteStatusConfigMap(kubeClient, namespace, api.ClusterAutoscalerStatus{AutoscalerStatus: "Initializing"}, nil, statusConfigMapName, time.Now())
```
I agree, thanks for the suggestion!

walidghallab commented:
Thanks a lot @MaciekPytel for the review!

/assign @MaciekPytel 
/cc @MaciekPytel 

Here is the output of the `kubectl get configmap cluster-autoscaler-status --namespace=kube-system -o yaml`
```yaml
apiVersion: v1
data:
  status: |
    time: 2023-12-20 10:52:04.558615784 +0000 UTC
    autoscalerStatus: Running
    clusterWide:
      health:
        status: Healthy
        nodeCounts:
          registered:
            total: 7
            ready: 7
            notStarted: 0
          longUnregistered: 0
          unregistered: 0
        lastProbeTime: "2023-12-20T10:52:04.558615784Z"
        lastTransitionTime: "2023-12-20T10:46:22.549253146Z"
      scaleUp:
        status: InProgress
        lastProbeTime: "2023-12-20T10:52:04.558615784Z"
        lastTransitionTime: "2023-12-20T10:52:04.558615784Z"
      scaleDown:
        status: NoCandidates
        lastProbeTime: "2023-12-20T10:51:54.408045905Z"
    nodeGroups:
    - name: https://www.googleapis.com/compute/v1/projects/sample-project/zones/us-central1-c/instanceGroups/gke-sample-cluster-default-pool-40ce0341-grp
      health:
        status: Healthy
        nodeCounts:
          registered:
            total: 7
            ready: 7
            notStarted: 0
          longUnregistered: 0
          unregistered: 0
        cloudProviderTarget: 10
        minSize: 4
        maxSize: 10
        lastProbeTime: "2023-12-20T10:52:04.558615784Z"
        lastTransitionTime: "2023-12-20T10:46:22.549253146Z"
      scaleUp:
        status: InProgress
        lastProbeTime: "2023-12-20T10:52:04.558615784Z"
        lastTransitionTime: "2023-12-20T10:52:04.558615784Z"
      scaleDown:
        status: NoCandidates
        lastProbeTime: "2023-12-20T10:51:54.408045905Z"
kind: ConfigMap
metadata:
  annotations:
    cluster-autoscaler.kubernetes.io/last-updated: 2023-12-20 10:52:04.558615784 +0000
      UTC
  creationTimestamp: "2023-12-20T10:46:10Z"
  name: cluster-autoscaler-status
  namespace: kube-system
  resourceVersion: "22049801"
  uid: 2be11b59-cf5c-4bd5-bb31-b7baf133bcbc
```

Output for `kubectl describe configmap cluster-autoscaler-status --namespace=kube-system`
```
Name:         cluster-autoscaler-status
Namespace:    kube-system
Labels:       <none>
Annotations:  cluster-autoscaler.kubernetes.io/last-updated: 2023-12-20 10:53:38.263729314 +0000 UTC

Data
====
status:
----
time: 2023-12-20 10:53:38.263729314 +0000 UTC
autoscalerStatus: Running
clusterWide:
  health:
    status: Healthy
    nodeCounts:
      registered:
        total: 7
        ready: 7
        notStarted: 0
      longUnregistered: 0
      unregistered: 0
    lastProbeTime: "2023-12-20T10:53:38.263729314Z"
    lastTransitionTime: "2023-12-20T10:46:22.549253146Z"
  scaleUp:
    status: NoActivity
    lastProbeTime: "2023-12-20T10:53:38.263729314Z"
    lastTransitionTime: "2023-12-20T10:52:45.533501545Z"
  scaleDown:
    status: NoCandidates
    lastProbeTime: "2023-12-20T10:53:38.263729314Z"
nodeGroups:
- name: https://www.googleapis.com/compute/v1/projects/sample-project/zones/us-central1-c/instanceGroups/gke-sample-cluster-default-pool-40ce0341-grp
  health:
    status: Healthy
    nodeCounts:
      registered:
        total: 7
        ready: 7
        notStarted: 0
      longUnregistered: 0
      unregistered: 0
    cloudProviderTarget: 7
    minSize: 4
    maxSize: 10
    lastProbeTime: "2023-12-20T10:53:38.263729314Z"
    lastTransitionTime: "2023-12-20T10:46:22.549253146Z"
  scaleUp:
    status: Backoff
    backoffInfo:
      errorCode: QUOTA_EXCEEDED
      errorMessage: 'Instance ''gke-sample-cluster-default-pool-40ce0341-xf5d'' creation
        failed: Quota ''CPUS'' exceeded.  Limit: 57.0 in region us-central1.'
    lastProbeTime: "2023-12-20T10:53:38.263729314Z"
    lastTransitionTime: "2023-12-20T10:52:45.533501545Z"
  scaleDown:
    status: NoCandidates
    lastProbeTime: "2023-12-20T10:53:38.263729314Z"


BinaryData
====

Events:
  Type     Reason         Age                   From                Message
  ----     ------         ----                  ----                -------
  Warning  ScaleUpFailed  7m13s                 cluster-autoscaler  Failed adding 1 nodes to group https://www.googleapis.com/compute/v1/projects/sample-project/zones/us-central1-c/instanceGroups/gke-sample-cluster-default-pool-40ce0341-grp due to OutOfResource.QUOTA_EXCEEDED; source errors: Instance 'gke-sample-cluster-default-pool-40ce0341-b9kr' creation failed: Quota 'CPUS' exceeded.  Limit: 57.0 in region us-central1.
  Warning  ScaleUpFailed  6m51s                 cluster-autoscaler  Failed adding 2 nodes to group https://www.googleapis.com/compute/v1/projects/sample-project/zones/us-central1-c/instanceGroups/gke-sample-cluster-default-pool-40ce0341-grp due to OutOfResource.QUOTA_EXCEEDED; source errors: Instance 'gke-sample-cluster-default-pool-40ce0341-dh9k' creation failed: Quota 'CPUS' exceeded.  Limit: 57.0 in region us-central1., Instance 'gke-sample-cluster-default-pool-40ce0341-swdz' creation failed: Quota 'CPUS' exceeded.  Limit: 57.0 in region us-central1.
  Normal   ScaledUpGroup  104s (x2 over 7m26s)  cluster-autoscaler  Scale-up: setting group https://www.googleapis.com/compute/v1/projects/sample-project/zones/us-central1-c/instanceGroups/gke-sample-cluster-default-pool-40ce0341-grp size to 10 instead of 7 (max: 10)
  Normal   ScaledUpGroup  101s (x2 over 7m23s)  cluster-autoscaler  Scale-up: group https://www.googleapis.com/compute/v1/projects/sample-project/zones/us-central1-c/instanceGroups/gke-sample-cluster-default-pool-40ce0341-grp size set to 10 instead of 7 (max: 10)
  Warning  ScaleUpFailed  91s                   cluster-autoscaler  Failed adding 1 nodes to group https://www.googleapis.com/compute/v1/projects/sample-project/zones/us-central1-c/instanceGroups/gke-sample-cluster-default-pool-40ce0341-grp due to OutOfResource.QUOTA_EXCEEDED; source errors: Instance 'gke-sample-cluster-default-pool-40ce0341-xm2m' creation failed: Quota 'CPUS' exceeded.  Limit: 57.0 in region us-central1.
  Warning  ScaleUpFailed  75s                   cluster-autoscaler  Failed adding 1 nodes to group https://www.googleapis.com/compute/v1/projects/sample-project/zones/us-central1-c/instanceGroups/gke-sample-cluster-default-pool-40ce0341-grp due to OutOfResource.QUOTA_EXCEEDED; source errors: Instance 'gke-sample-cluster-default-pool-40ce0341-vgbp' creation failed: Quota 'CPUS' exceeded.  Limit: 57.0 in region us-central1.
  Warning  ScaleUpFailed  63s                   cluster-autoscaler  Failed adding 1 nodes to group https://www.googleapis.com/compute/v1/projects/sample-project/zones/us-central1-c/instanceGroups/gke-sample-cluster-default-pool-40ce0341-grp due to OutOfResource.QUOTA_EXCEEDED; source errors: Instance 'gke-sample-cluster-default-pool-40ce0341-xf5d' creation failed: Quota 'CPUS' exceeded.  Limit: 57.0 in region us-central1.
```

I changed the project name in the messages above.

I will refactor config map tests in a subsequent PR since it is out of scope of this PR.

walidghallab commented on cluster-autoscaler/clusterstate/api/types.go:
```diff
@@ -69,36 +69,148 @@ const (
+	NodeCounts NodeGroupScaleUpNodeCount `json:"nodeCounts,omitempty" yaml:"nodeCounts,omitempty"`
+	// LastProbeTime is the last time we probed the condition.
+	BackoffInfo   BackoffInfo `json:"backoffInfo,omitempty" yaml:"backoffInfo,omitempty"`
+	LastProbeTime metav1.Time `json:"lastProbeTime,omitempty" yaml:"lastProbeTime,omitempty"`
```
Fixed

walidghallab commented on cluster-autoscaler/clusterstate/clusterstate.go:
```diff
@@ -802,35 +804,47 @@ func (csr *ClusterStateRegistry) GetClusterReadiness() Readiness {
-			acceptable.CurrentTarget),
+func (csr *ClusterStateRegistry) buildScaleUpStatusNodeGroup(nodeGroup cloudprovider.NodeGroup, readiness Readiness, acceptable AcceptableRange, now time.Time, lastStatus api.NodeGroupScaleUpCondition) api.NodeGroupScaleUpCondition {
+	isScaleUpInProgress := csr.IsNodeGroupScalingUp(nodeGroup.Id())
+	scaleUpSafety := csr.IsNodeGroupSafeToScaleUp(nodeGroup, now)
```
Good point, renamed. Thank you!

walidghallab commented:
/hold Holding submission to make sure [this comment](https://github.com/kubernetes/autoscaler/issues/6318#issuecomment-1847696829) is addressed before the bot submit automatically.

walidghallab commented:
/unhold the comment has been addressed.
I've added a new field `beingDeleted` and added tag `omitDefault` to it to not display it when it is 0 (which happens most of the time).

walidghallab commented:
/unhold

MaciekPytel commented on cluster-autoscaler/clusterstate/api/types.go:
```diff
@@ -69,36 +64,135 @@ const (
+	LastProbeTime metav1.Time `json:"lastProbeTime,omitempty" yaml:"lastProbeTime,omitempty"`
 	// LastTransitionTime is the time since when the condition was in the given state.
-	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty" yaml:"lastTransitionTime,omitempty"`
```
Again, this is not a comment to this PR as some thought for potential follow-up: do we actually need those probe/transition times? What we used previously was essentially standard Kubernetes status conditions, so naturally they had a standard set of fields.

New structs look very differently from kube status conditions, so there is no point adhering to them as standard. And I'm thinking maybe those fields are not adding much value:
1. We define lastTransitionTime as time when Status field changed. But that seems rather arbitrary, given that we don't consider change in other fields as transition.
2. If we got rid of timestamps maybe we wouldn't need to write the configmap every loop? We could just write-on-change which should save constant writes to api-server in most clusters (admittedly that saves QPS when clusters are likely not under heavy load and doesn't help at all when nodes are being added - which is most likely when api-server is experiencing maximum load).
3. We could get rid of extra logic needed to handle those fields.

WDYT @x13n @walidghallab ?

MaciekPytel commented on cluster-autoscaler/clusterstate/clusterstate.go:
```diff
@@ -839,130 +846,114 @@ func buildScaleUpStatusNodeGroup(isScaleUpInProgress bool, scaleUpSafety NodeGro
 		condition.Status = api.ClusterAutoscalerBackoff
+		condition.BackoffInfo = api.BackoffInfo{
+			ErrorCode:    scaleUpSafety.BackoffStatus.ErrorInfo.ErrorCode,
+			ErrorMessage: scaleUpSafety.BackoffStatus.ErrorInfo.ErrorMessage,
```
I wonder if we should maybe trim this to some maximum length? Max size of configmap is 1MB and that's not much if you have a few hundred nodegroups. 
A few hundred characters should likely be a reasonable limit for an error message?

walidghallab replied:
I actually don't print all the error messages. Only the first three error messages in each node group are printed.
They are also the same error messages used in events.

I added truncation logic for extra safety, PTAL.

MaciekPytel commented on cluster-autoscaler/clusterstate/clusterstate.go:
```diff
@@ -839,130 +846,114 @@ func buildScaleUpStatusNodeGroup(isScaleUpInProgress bool, scaleUpSafety NodeGro
-		),
+func buildHealthStatusClusterwide(isHealthy bool, readiness Readiness, lastStatus api.ClusterHealthCondition) api.ClusterHealthCondition {
+	condition := api.ClusterHealthCondition{
+		NodeCounts: api.NodeCount{
```
nit: This happens at least twice, maybe it would be worth having a function `NodeCountsFromReadiness(readiness Readiness) api.NodeCount`? It's less boilerplate and more importantly it makes it easier to do changes - like the one to representing unregistered that I mentioned above.

walidghallab replied:
Done.

MaciekPytel commented on cluster-autoscaler/processors/actionablecluster/actionable_cluster_processor.go:
```diff
@@ -71,7 +72,7 @@ func OnEmptyCluster(context *context.AutoscalingContext, status string, emitEven
 	metrics.UpdateNodesCount(0, 0, 0, 0, 0)
 	if context.WriteStatusConfigMap {
-		utils.WriteStatusConfigMap(context.ClientSet, context.ConfigNamespace, status, context.LogRecorder, context.StatusConfigMapName)
+		utils.WriteStatusConfigMap(context.ClientSet, context.ConfigNamespace, api.ClusterAutoscalerStatus{AutoscalerStatus: "Initializing", Message: status}, context.LogRecorder, context.StatusConfigMapName, time.Now())
```
use helpers / consts here as well

MaciekPytel reviewed (commented):
/lgtm
/approve
I left a bunch of non-blocking comments for a potential discussion / follow-up. Let's not fix them on this PR.

/hold
For fixing one comment that I see as blocking - the one about using "Initializing" instead of const in actionable_cluster_processor. Please feel free to remove hold as soon as you address this one.

MaciekPytel commented:
/lgtm
/hold cancel
